From a0d6815f298dc09f65e1a03f8021ec63cbd01d8a Mon Sep 17 00:00:00 2001
From: Dmitri Epshtein <dima@marvell.com>
Date: Mon, 24 Jul 2017 13:29:04 +0300
Subject: [PATCH 1953/2241] neta: improve performance in HWBM when GRO disable

use netif_receive_skb() instead of napi_gro_receive()
if GRO is disabled (as for SWBM)

Change-Id: I0105fb73813e84ff75d34b217f84b5ceb77efe25
Signed-off-by: Dmitri Epshtein <dima@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/42051
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Neta Zur Hershkovits <neta@marvell.com>
---
 drivers/net/ethernet/marvell/mvneta.c | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/drivers/net/ethernet/marvell/mvneta.c b/drivers/net/ethernet/marvell/mvneta.c
index 8fe4747..0f341be 100644
--- a/drivers/net/ethernet/marvell/mvneta.c
+++ b/drivers/net/ethernet/marvell/mvneta.c
@@ -2350,7 +2350,10 @@ err_drop_frame:
 
 			skb->protocol = eth_type_trans(skb, dev);
 			mvneta_rx_csum(pp, rx_status, skb);
-			napi_gro_receive(napi, skb);
+			if (dev->features & NETIF_F_GRO)
+				napi_gro_receive(napi, skb);
+			else
+				netif_receive_skb(skb);
 
 			rcvd_pkts++;
 			rcvd_bytes += rx_bytes;
@@ -2393,7 +2396,10 @@ err_drop_frame:
 
 		mvneta_rx_csum(pp, rx_status, skb);
 
-		napi_gro_receive(napi, skb);
+		if (dev->features & NETIF_F_GRO)
+			napi_gro_receive(napi, skb);
+		else
+			netif_receive_skb(skb);
 	}
 
 	if (rcvd_pkts) {
-- 
2.7.4

